### PR TITLE
Us80722 - Visual cues for inactive courses

### PR DESCRIPTION
**Pull request by jstefaniuk-d2l**
You may need a LP version recent to Jan 12

Add in visual cues notifying users of inactive courses

image

Story:
https://rally1.rallydev.com/#/51191528503d/detail/userstory/81532795136?fdp=true

Acceptance criteria:
- If the course has ended, the inactive state does not matter, show the "Course Ended" overlay as it would normally
- If the course starting soon and is inactive, add "(Inactive)" under the date
- If the course starting soon and is active, do not add "(Inactive)" under the date
- If the course is inactive, and has no starting date, add an overlay with the text "Inactive"
- If the course has started, but is inactive:
    - Have an overlay with "Course Started" in bold and "(Inactive)" underneath
    - Place a yellow indicator at the top right of the course tile
    - Display an alert in the my-courses page notifying the user
        - This alert should appear/disappear appropriately when things are pinned/unpinned
             - From both the navbar and widget
        - Having 2 started inactive courses and unpinning one should not remove the alert

Risky areas:
- Pinning/Unpinning things in different ways

### CODE REVIEW

File: d2l-course-tile-grid.html [hunk 1]
```diff
@@ -446,6 +446,18 @@
 
 					this._courseTileBoundingRects.push(offsetData);
 				}.bind(this));
+			},
+			checkForStartedInactive: function(type) {
+				var courseTiles = this.$$('.course-tile-container').querySelectorAll('d2l-course-tile');
+				var searchAmount = type === 'remove' ? 2 : 1;
```
Comment (CodeBaboon) [inline d2l-course-tile-grid.html]: This had me puzzled until talking to Jeff. Turns out the tile won't actually be gone yet, hence the need for two checks. Probably could use a comment.

**Comment by jstefaniuk-d2l** [comment]
- [ ] unit tests

File: test/d2l-course-tile/d2l-course-tile.js [hunk 2]
```diff
@@ -464,4 +464,139 @@ describe('<d2l-course-tile>', function() {
...
+
+		beforeEach(function() {
+			org = {
+				properties: {
+					endDate: getFutureDate(),
+					startDate: getPastDate(),
+					isActive: true
+				}
+			};
+			console.log('org', org);
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: no console pls

File: test/d2l-course-tile/d2l-course-tile.js [hunk 3]
```diff
@@ -464,4 +464,139 @@ describe('<d2l-course-tile>', function() {
...
+			};
+			window.BSI = window.BSI || {};
+			window.BSI.Intl = window.BSI.Intl || {
+				DateTimeFormat: function() {
+					this.format = sinon.stub().returns(formattedDate);
+				}
+			};
+		});
+
+		describe('course not started', function() {
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: all these describes could be a little more englishy, or even ideally more [Gherkiny](https://github.com/cucumber/cucumber/wiki/Gherkin). 

```
describe('given the notification overlay is opened')
   describe('when the course is not started')
      describe('and the course is active then')
        ('it adds an overlay with the course start date')
```

File: test/d2l-course-tile/d2l-course-tile.js [hunk 4]
```diff
@@ -464,4 +464,139 @@ describe('<d2l-course-tile>', function() {
...
+		if (date) {
+			expect(overlayDate.textContent).to.equal(formattedDate)
+		} else {
+			expect(overlayDate.textContent).to.not.equal(formattedDate)
+		}
+
+		if (inactive) {
+			expect(overlayInactive.textContent).to.equal(inactiveText)
+		} else {
+			expect(overlayDate.textContent).to.not.equal(inactiveText)
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: copy pasta, should be `overlayInactive`

File: test/d2l-course-tile/d2l-course-tile.js [hunk 5]
```diff
@@ -464,4 +464,139 @@ describe('<d2l-course-tile>', function() {
...
+					verifyOverlay('Course Ended', true, false);
+				});
+			});
+
+			describe('course inactive', function() {
+				it('Adds an overlay with the date', function() {
+					org.properties.endDate = getPastDate();
+					org.properties.isActive = false;
+					widget._checkDateBounds(org, response);
+					verifyOverlay('Course Ended', true, false);
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: Shouldn't this be `verifyOverlay('Course Ended', true, true);` ?

File: test/d2l-course-tile/d2l-course-tile.js [hunk 6]
```diff
@@ -464,4 +464,139 @@ describe('<d2l-course-tile>', function() {
...
+	}
+
+	function getCurrentDate() {
+		return new Date(curDate).toISOString()
+	}
+
+	var formattedDate = 'FORMATTED_DATE';
+	var inactiveText = '(Inactive)';
+
+	function verifyOverlay(title, date, inactive) {
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: This convenience function actually makes the tests themselves harder to understand in terms of their expectations. Not sure it is worth it.

File: test/d2l-course-tile/d2l-course-tile.js [hunk 7]
```diff
@@ -464,4 +464,139 @@ describe('<d2l-course-tile>', function() {
...
+					verifyOverlay('', false, false);
+				});
+			});
+
+			describe('course inactive', function() {
+				it('adds an "inactive" overlay', function() {
+					org.properties.startDate = null;
+					org.properties.isActive = false;
+					widget._checkDateBounds(org, response);
+					verifyOverlay('Inactive', false, false);
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: Should this be `verifyOverlay('Inactive', false, true);` ?

File: test/d2l-course-tile/d2l-course-tile.js [hunk 8]
```diff
@@ -486,15 +486,15 @@ describe('<d2l-course-tile>', function() {
...
-			expect(overlayDate.textContent).to.not.equal(formattedDate)
+			expect(overlayDate.textContent).to.not.equal(formattedDate);
 		}
 
 		if (inactive) {
-			expect(overlayInactive.textContent).to.equal(inactiveText)
+			expect(overlayInactive.textContent).to.equal(inactiveText);
 		} else {
-			expect(overlayDate.textContent).to.not.equal(inactiveText)
+			expect(overlayDate.textContent).to.not.equal(inactiveText);
```
Comment (CodeBaboon) [inline test/d2l-course-tile/d2l-course-tile.js]: copy pasta, should be `overlayInactive` ?

File: test/d2l-course-tile/d2l-course-tile.js (hunk 5 above)
Comment (jstefaniuk-d2l) [inline test/d2l-course-tile/d2l-course-tile.js]: Course ended shows no inactive indicator

File: test/d2l-course-tile/d2l-course-tile.js (hunk 7 above)
Comment (jstefaniuk-d2l) [inline test/d2l-course-tile/d2l-course-tile.js]: Inactive is the title text, the "(inactive)" indicator is blank here

File: test/d2l-course-tile/d2l-course-tile.js (hunk 3 above)
Comment (jstefaniuk-d2l) [inline test/d2l-course-tile/d2l-course-tile.js]: - [ ] this

**Comment by ctwomey1** [comment]
Testing Checklist per browser:
Chrome:
- [ ] If the course has ended, the inactive state does not matter, show the "Course Ended" overlay as it would normally
- [ ] If the course starting soon and is inactive, add "(Inactive)" under the date
- [ ] If the course starting soon and is active, do not add "(Inactive)" under the date
- [ ] If the course is inactive, and has no starting date, add an overlay with the text "Inactive"
- [ ] If the course has started, but is inactive:
        Have an overlay with "Course Started" in bold and "(Inactive)" underneath
        Place a yellow indicator at the top right of the course tile
        Display an alert in the my-courses page notifying the user
        This alert should appear/disappear appropriately when things are pinned/unpinned From both the navbar and widget

- [ ] Having 2 started inactive courses and unpinning one should not remove the alert

Safari:
- [ ] If the course has ended, the inactive state does not matter, show the "Course Ended" overlay as it would normally
- [ ] If the course starting soon and is inactive, add "(Inactive)" under the date
- [ ] If the course starting soon and is active, do not add "(Inactive)" under the date
- [ ] If the course is inactive, and has no starting date, add an overlay with the text "Inactive"
- [ ] If the course has started, but is inactive:
        Have an overlay with "Course Started" in bold and "(Inactive)" underneath
        Place a yellow indicator at the top right of the course tile
        Display an alert in the my-courses page notifying the user
        This alert should appear/disappear appropriately when things are pinned/unpinned From both the navbar and widget

- [ ] Having 2 started inactive courses and unpinning one should not remove the alert

FireFox:
- [ ] If the course has ended, the inactive state does not matter, show the "Course Ended" overlay as it would normally
- [ ] If the course starting soon and is inactive, add "(Inactive)" under the date
- [ ] If the course starting soon and is active, do not add "(Inactive)" under the date
- [ ] If the course is inactive, and has no starting date, add an overlay with the text "Inactive"
- [ ] If the course has started, but is inactive:
        Have an overlay with "Course Started" in bold and "(Inactive)" underneath
        Place a yellow indicator at the top right of the course tile
        Display an alert in the my-courses page notifying the user
        This alert should appear/disappear appropriately when things are pinned/unpinned From both the navbar and widget

- [ ] Having 2 started inactive courses and unpinning one should not remove the alert

Edge:
- [ ] If the course has ended, the inactive state does not matter, show the "Course Ended" overlay as it would normally
- [ ] If the course starting soon and is inactive, add "(Inactive)" under the date
- [ ] If the course starting soon and is active, do not add "(Inactive)" under the date
- [ ] If the course is inactive, and has no starting date, add an overlay with the text "Inactive"
- [ ] If the course has started, but is inactive:
        Have an overlay with "Course Started" in bold and "(Inactive)" underneath
        Place a yellow indicator at the top right of the course tile
        Display an alert in the my-courses page notifying the user
        This alert should appear/disappear appropriately when things are pinned/unpinned From both the navbar and widget

- [ ] Having 2 started inactive courses and unpinning one should not remove the alert


ie:
- [ ] If the course has ended, the inactive state does not matter, show the "Course Ended" overlay as it would normally
- [ ] If the course starting soon and is inactive, add "(Inactive)" under the date
- [ ] If the course starting soon and is active, do not add "(Inactive)" under the date
- [ ] If the course is inactive, and has no starting date, add an overlay with the text "Inactive"
- [ ] If the course has started, but is inactive:
        Have an overlay with "Course Started" in bold and "(Inactive)" underneath
        Place a yellow indicator at the top right of the course tile
        Display an alert in the my-courses page notifying the user
        This alert should appear/disappear appropriately when things are pinned/unpinned From both the navbar and widget

- [ ] Having 2 started inactive courses and unpinning one should not remove the alert

**Comment by ctwomey1** [comment]
Tested in above browsers successfully, it appears to work fine across all browsers.

**Pull request merged**